# Introduction
⁃	Q1: Load the dataset.

In [1]:
import pandas as pd
df = pd.read_csv('book_data.csv')
df

book_authors  \
0                                    Kristin Hannah   
1                                   Madeline Miller   
2                                       Holly Black   
3                                         A.J. Finn   
4                                     Tara Westover   
5                                      Tomi Adeyemi   
6                                      Tayari Jones   
7     Michelle McNamara|Gillian Flynn|Patton Oswalt   
8                    Greer Hendricks|Sarah Pekkanen   
9                                   Neal Shusterman   
10                                    Sarah J. Maas   
11                                  Chloe  Benjamin   
12                                     Stephen King   
13                          Helen Hoang|Helen Hoang   
14                                      Delia Owens   
15                                  Shantel Tessier   
16                                   Adrienne Young   
17                                        Ruth Ware   
18                                      Naomi Novik   
19                        Amie Kaufman|Jay Kristoff   
20                                  Shantel Tessier   
21                                 Simone St. James   
22                                       Jojo Moyes   
23                                       Vi Keeland   
24                                 Heather   Morris   
25                                     Pierce Brown   
26                                 Becky Albertalli   
27                                      Sabaa Tahir   
28                                   Melissa Albert   
29                                   Colleen Hoover   
...                                             ...   
1111                                  Makiia Lucier   
1112                                 Shannon Stacey   
1113                                Waubgeshig Rice   
1114                                 Mariana Zapata   
1115                           Tamara Ireland Stone   
1116                                   Caitlin Macy   
1117                                 Seanan McGuire   
1118                                      J.R. Ward   
1119                                    Dean Koontz   
1120                           Stefan Merrill Block   
1121                            Susan Wittig Albert   
1122                                 Colleen Hoover   
1123                                 Preston Norton   
1124                   Bill Clinton|James Patterson   
1125                              Sam Graham-Felsen   
1126                  Douglas Preston|Lincoln Child   
1127                                     Kasie West   
1128                                         Kim Fu   
1129                                 Mariana Zapata   
1130                               Christine Higdon   
1131            Laurie Halse Anderson|Emily Carroll   
1132                                   Ava Dellaira   
1133                                Debbie  Johnson   
1134                                  Abigail Strom   
1135                                   Lexa Hillyer   
1136                                 Alyssa Brandon   
1137                               Crystal Hana Kim   
1138                                  Leslie  Cohen   
1139                               Sarah Beth Durst   
1140                                   Nisha Sharma   

                                              book_desc    book_edition  \
0     Alaska, 1974.Unpredictable. Unforgiving. Untam...             NaN   
1     In the house of Helios, god of the sun and mig...             NaN   
2     Of course I want to be like them. They’re beau...             NaN   
3     Anna Fox lives alone—a recluse in her New York...             NaN   
4     An unforgettable memoir in the tradition of Th...             NaN   
5     They killed my mother.They took our magic.They...             NaN   
6     Newlyweds Celestial and Roy are the embodiment...             NaN   
7     A masterful true crime

⁃	Q2: How many books are there? 

In [2]:
df.shape[0]

1141

⁃	Q3: How many books have more than 1 author?

In [3]:
df.loc[df['book_authors'].str.contains('\|')].shape[0]

144

# ⁃	Length:
⁃	Q4: You will want to create a new column with the integer value of the number of pages.(If you remove rows in this process, please state why.) 

I add a new column called "pages_num" as the integer value of the number of pages. Since NaN cannot be transferred to int, I use -1 instead of it to represent NaN.

In [4]:
pages_num = df['book_pages'].copy()
pages_num = pages_num.str.replace(" pages",'').fillna(-1).astype(int)
df['pages_num'] = pages_num
df

book_authors  \
0                                    Kristin Hannah   
1                                   Madeline Miller   
2                                       Holly Black   
3                                         A.J. Finn   
4                                     Tara Westover   
5                                      Tomi Adeyemi   
6                                      Tayari Jones   
7     Michelle McNamara|Gillian Flynn|Patton Oswalt   
8                    Greer Hendricks|Sarah Pekkanen   
9                                   Neal Shusterman   
10                                    Sarah J. Maas   
11                                  Chloe  Benjamin   
12                                     Stephen King   
13                          Helen Hoang|Helen Hoang   
14                                      Delia Owens   
15                                  Shantel Tessier   
16                                   Adrienne Young   
17                                        Ruth Ware   
18                                      Naomi Novik   
19                        Amie Kaufman|Jay Kristoff   
20                                  Shantel Tessier   
21                                 Simone St. James   
22                                       Jojo Moyes   
23                                       Vi Keeland   
24                                 Heather   Morris   
25                                     Pierce Brown   
26                                 Becky Albertalli   
27                                      Sabaa Tahir   
28                                   Melissa Albert   
29                                   Colleen Hoover   
...                                             ...   
1111                                  Makiia Lucier   
1112                                 Shannon Stacey   
1113                                Waubgeshig Rice   
1114                                 Mariana Zapata   
1115                           Tamara Ireland Stone   
1116                                   Caitlin Macy   
1117                                 Seanan McGuire   
1118                                      J.R. Ward   
1119                                    Dean Koontz   
1120                           Stefan Merrill Block   
1121                            Susan Wittig Albert   
1122                                 Colleen Hoover   
1123                                 Preston Norton   
1124                   Bill Clinton|James Patterson   
1125                              Sam Graham-Felsen   
1126                  Douglas Preston|Lincoln Child   
1127                                     Kasie West   
1128                                         Kim Fu   
1129                                 Mariana Zapata   
1130                               Christine Higdon   
1131            Laurie Halse Anderson|Emily Carroll   
1132                                   Ava Dellaira   
1133                                Debbie  Johnson   
1134                                  Abigail Strom   
1135                                   Lexa Hillyer   
1136                                 Alyssa Brandon   
1137                               Crystal Hana Kim   
1138                                  Leslie  Cohen   
1139                               Sarah Beth Durst   
1140                                   Nisha Sharma   

                                              book_desc    book_edition  \
0     Alaska, 1974.Unpredictable. Unforgiving. Untam...             NaN   
1     In the house of Helios, god of the sun and mig...             NaN   
2     Of course I want to be like them. They’re beau...             NaN   
3     Anna Fox lives alone—a recluse in her New York...             NaN   
4     An unforgettable memoir in the tradition of Th...             NaN   
5     They killed my mother.They took our magic.They...             NaN   
6     Newlyweds Celestial and Roy are the embodiment...             NaN   
7     A masterful true crime

⁃	Q5: What is the average number of pages?

In [5]:
df[df['pages_num']>=0]['pages_num'].mean()

327.9000933706816

⁃	Q6: What is the median number of pages? 

In [6]:
df[df['pages_num']>=0]['pages_num'].median()

328.0

⁃	Q7: Does having more than 1 author result in a longer book on average? What is the average number of pages for books written by a single author? What is it for when there are at least two authors? (We will do more careful analysis for these types of questions later. For now, we just want you to practice using some DataFrame functionalities). 

In [7]:
df_single = df.loc[~df['book_authors'].str.contains('\|')]
df_single[df_single['pages_num']>=0]['pages_num'].mean()

334.2275641025641

In [8]:
df_multiple = df.loc[df['book_authors'].str.contains('\|')]
df_multiple[df_multiple['pages_num']>=0]['pages_num'].mean()

284.02962962962965

The average number of pages for books written by a single author is 334.2275641025641. And the average number of pages for books written by at least two authors is 284.02962962962965. So having more than 1 author doesn't result in a longer book on average.

# Ratings
⁃	Q8: How many books have at least a rating of 4.25? 

In [9]:
df[df['book_rating'] >= 4.25].shape[0]

305

⁃	Q9: How about at most a rating of 4?

In [10]:
df[df['book_rating'] <= 4].shape[0]

494

⁃	Q10: Discretize the ratings. (1,2,3,4,5). 

In [11]:
pd.cut(x = df['book_rating'], bins = [1,2,3,4,5])

0       (4, 5]
1       (4, 5]
2       (4, 5]
3       (3, 4]
4       (4, 5]
5       (4, 5]
6       (4, 5]
7       (4, 5]
8       (3, 4]
9       (4, 5]
10      (3, 4]
11      (3, 4]
12      (4, 5]
13      (3, 4]
14      (4, 5]
15      (4, 5]
16      (4, 5]
17      (3, 4]
18      (4, 5]
19      (4, 5]
20      (4, 5]
21      (4, 5]
22      (4, 5]
23      (4, 5]
24      (4, 5]
25      (4, 5]
26      (4, 5]
27      (4, 5]
28      (3, 4]
29      (4, 5]
         ...  
1111    (3, 4]
1112    (4, 5]
1113    (4, 5]
1114    (4, 5]
1115    (3, 4]
1116    (2, 3]
1117    (4, 5]
1118    (4, 5]
1119    (4, 5]
1120    (3, 4]
1121    (4, 5]
1122    (4, 5]
1123    (4, 5]
1124    (3, 4]
1125    (3, 4]
1126    (4, 5]
1127    (3, 4]
1128    (3, 4]
1129    (4, 5]
1130    (4, 5]
1131    (4, 5]
1132    (4, 5]
1133    (4, 5]
1134    (3, 4]
1135    (3, 4]
1136    (3, 4]
1137    (3, 4]
1138    (3, 4]
1139    (3, 4]
1140    (3, 4]
Name: book_rating, Length: 1141, dtype: category
Categories (4, interval[int64]): [(1

⁃	Q11: For each of the discretized ratings (1,2,3,4,5), what is the average number of pages?

In [12]:
df_countablepages = df[df['pages_num'] >= 0]
df_countablepages.groupby(pd.cut(x = df_countablepages['book_rating'], bins = [1,2,3,4,5]))['pages_num'].mean()

book_rating
(1, 2]           NaN
(2, 3]    285.333333
(3, 4]    319.407173
(4, 5]    334.892256
Name: pages_num, dtype: float64

⁃	Q12: How about the average number of reviews?

In [13]:
df.groupby(pd.cut(x = df['book_rating'], bins = [1,2,3,4,5]))['book_review_count'].mean()

book_rating
(1, 2]           NaN
(2, 3]    251.333333
(3, 4]    911.213849
(4, 5]    857.219474
Name: book_review_count, dtype: float64

# Genres:
⁃	Q13: Create a new DataFrame, exploding the rows with multiple genres such that it is one row per genre/book.

In [14]:
genres = df['genres'].str.split('|',expand = True).stack()
genres = genres.reset_index(level=1,drop=True).rename('genre')
df2 = df.copy()
df2 = df2.join(genres)
df2

book_authors                                          book_desc  \
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
2          Holly Black  Of course I want to be like them. They’re beau...   
2          Holly Black  Of course I want to be like them. They’re beau...   
2          Holly Black  Of course I want to be like them. They’re beau...   
2          Holly Black  Of course I want to be like them. They’re beau...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
...                ...                                                ...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1138     Leslie  Cohen  This is the classic tale of boy meets girl: Gi...   
1138     Leslie  Cohen  This is the classic tale of boy meets girl: Gi...   
1138     Leslie  Cohen  This is the classic tale of boy meets girl: Gi...   
1138     Leslie  Cohen  This is the classic tale of boy meets girl: Gi...   
1138     Leslie  Cohen  This is the classic tale of boy meets girl: Gi...   
1138  

⁃	Q14: What is the average rating of books from different genres? What is the median?

In [15]:
df2.groupby(df2['genre'])['book_rating'].mean()

genre
16th Century                      4.130000
20th Century                      4.000000
Abandoned                         3.557692
Abuse                             4.365000
Academic                          4.193333
Activism                          4.200000
Adult                             3.951148
Adult Fiction                     3.967451
Adventure                         4.120769
Africa                            4.018000
African American                  3.970000
Aliens                            4.060000
Alternate History                 3.855000
Amazon                            4.275000
American                          4.030000
American History                  4.133000
Amish                             4.240000
Angels                            4.485000
Animals                           4.044286
Anthologies                       3.943333
Apocalyptic                       4.196000
Art                               3.933333
Asia                              3.920000
Asian

In [16]:
df2.groupby(df2['genre'])['book_rating'].median()

genre
16th Century                      4.130
20th Century                      4.000
Abandoned                         3.630
Abuse                             4.300
Academic                          4.130
Activism                          4.200
Adult                             3.970
Adult Fiction                     3.940
Adventure                         4.080
Africa                            4.030
African American                  3.990
Aliens                            4.060
Alternate History                 3.755
Amazon                            4.275
American                          4.040
American History                  4.270
Amish                             4.240
Angels                            4.485
Animals                           4.030
Anthologies                       3.880
Apocalyptic                       4.060
Art                               3.875
Asia                              3.830
Asian Literature                  3.955
Atheism                           

⁃	Q15: What is the average length of books from different genres? What is the median?

In [17]:
df2_countablepages = df2[df2['pages_num']>=0]
df2_countablepages.groupby(df2_countablepages['genre'])['pages_num'].mean()

genre
16th Century                      560.000000
20th Century                      256.000000
Abandoned                         343.923077
Abuse                             359.000000
Academic                          335.833333
Activism                          166.000000
Adult                             359.016393
Adult Fiction                     294.765957
Adventure                         334.000000
Africa                            274.600000
African American                  262.750000
Alternate History                 283.000000
Amazon                            334.000000
American                          284.000000
American History                  444.111111
Amish                             308.000000
Angels                            569.000000
Animals                           278.428571
Anthologies                       399.666667
Apocalyptic                       530.000000
Art                               308.200000
Asia                              348.000000
Asia

In [18]:
df2_countablepages.groupby(df2_countablepages['genre'])['pages_num'].median()

genre
16th Century                      560.0
20th Century                      256.0
Abandoned                         308.0
Abuse                             349.0
Academic                          329.5
Activism                          166.0
Adult                             336.0
Adult Fiction                     304.0
Adventure                         340.0
Africa                            288.0
African American                  264.0
Alternate History                 328.0
Amazon                            334.0
American                          280.0
American History                  402.0
Amish                             308.0
Angels                            569.0
Animals                           304.0
Anthologies                       368.0
Apocalyptic                       501.0
Art                               336.0
Asia                              352.0
Asian Literature                  269.5
Atheism                           176.0
Audiobook                         